### PR TITLE
Added modelled exception on DIG

### PR DESCRIPTION
**Pull request by minsunny**
https://github.com/nanoPayinc/NANOPAY/issues/2688

Created error models, to deal with every DIG error case with an internal code starting from 1000.
It is shown with error code, type, message etc. as required format.

### CODE REVIEW

File: src/foam/nanos/dig/DigWebAgent.java [hunk 1]
```diff
@@ -96,8 +97,11 @@ public void execute(X x) {
       DAO dao = (DAO) x.get(daoName);
 
       if ( dao == null ) {
-        resp.sendError(HttpServletResponse.SC_NOT_FOUND, "Unknown DAO: "+daoName);
-        throw new RuntimeException("DAO not found");
+        DigErrorMessage error = new DaoNoFoundException.Builder(x)
+                                      .setMessage("DAO can not found: " + daoName)
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: replace with:  DAO not found:

File: src/foam/nanos/dig/DigWebAgent.java [hunk 2]
```diff
@@ -120,64 +124,47 @@ public void execute(X x) {
           outputterJson.setOutputClassNames(false);
           // let FObjectArray parse first
           if ( SafetyUtil.isEmpty(data) ) {
-              resp.sendError(HttpServletResponse.SC_BAD_REQUEST, "PUT|POST expecting data, non received.");
+            //TODO: return exception
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: remove.

File: src/foam/nanos/dig/DigWebAgent.java [hunk 3]
```diff
@@ -120,64 +124,47 @@ public void execute(X x) {
...
+              outputException(x, resp, format, out, error);
               return;
           }
           try {
             Object o = jsonParser.parseStringForArray(data, objClass);
-            if ( o != null && o instanceof Object[] ) {
+            Object o1 = jsonParser.parseString(data, objClass);
+            if ( o == null && o1 == null ) {
+              DigErrorMessage error = new ParsingErrorException.Builder(x)
+                                            .setMessage("Invalid Json Format")
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: replace with: Invalid JSON Format

File: src/foam/nanos/dig/DigWebAgent.java [hunk 4]
```diff
@@ -212,20 +203,29 @@ public void execute(X x) {
...
             return;
           }
 
           for ( int i = 0 ; i < list.size() ; i++ ) {
             dao.put((FObject) list.get(i));
           }
         } else if ( Format.HTML == format ) {
-          resp.sendError(HttpServletResponse.SC_NOT_ACCEPTABLE, "Unsupported Accept");
+          DigErrorMessage error = new UnsupportException.Builder(x)
+                                        .setMessage("This" + format + "is not supported.")
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: replace with: Unsupported Format:   and provide the format string if possible.

File: src/foam/nanos/dig/DigWebAgent.java [hunk 5]
```diff
@@ -327,8 +327,12 @@ public void execute(X x) {
           if ( Format.XML == format ) {
             resp.setContentType("text/html");
           }
-          out.println("unsuported DAO");
-          resp.setStatus(HttpServletResponse.SC_OK);
+
+          DigErrorMessage error = new ParsingErrorException.Builder(x)
+            .setMessage("Unsupported DAO")
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: Provide the dao name if possible.

File: src/foam/nanos/dig/exception/DaoNoFoundException.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.dig.exception',
+  name: 'DaoNoFoundException',
```
Comment (jlhughes) [inline src/foam/nanos/dig/exception/DaoNoFoundException.js]: Rename to DAONotFoundException

File: src/foam/nanos/dig/exception/DaoNoFoundException.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    },
+    {
+      class: 'String',
+      name: 'code',
+      value: '1000'
+    },
+    {
+      class: 'String',
+      name: 'type',
+      value: 'NoFound'
```
Comment (jlhughes) [inline src/foam/nanos/dig/exception/DaoNoFoundException.js]: Replace with: NotFound

File: src/foam/nanos/dig/exception/DaoPutException.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/**ß
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.dig.exception',
+  name: 'DaoPutException',
```
Comment (jlhughes) [inline src/foam/nanos/dig/exception/DaoPutException.js]: Rename to DAOPutException

**Review by jlhughes — CHANGES_REQUESTED** [review]

File: src/foam/lib/csv/CSVSupport.java [hunk 9]
```diff
@@ -43,7 +43,10 @@ public void inputCSV(InputStream is, Sink sink, ClassInfo classInfo) {
       //get propertyInfo from ClassInfo
       for ( int i = 0 ; i < column ; i++ ) {
         propertyInfos[i] = (PropertyInfo) classInfo.getAxiomByName((String) propNames[i]);
+        if ( propertyInfos[i] == null )
+          throw new RuntimeException("CSV head format error");
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSVSupport.java]: What is a CSV head format error?

File: src/foam/nanos/dig/DigWebAgent.java [hunk 10]
```diff
@@ -60,7 +61,7 @@ public void execute(X x) {
     Logger              logger   = (Logger) x.get("logger");
     HttpServletResponse resp     = x.get(HttpServletResponse.class);
     HttpParameters      p        = x.get(HttpParameters.class);
-    final PrintWriter   out      = x.get(PrintWriter.class);
+    PrintWriter   out            = x.get(PrintWriter.class);
```
Comment (kgrgreer) [inline src/foam/nanos/dig/DigWebAgent.java]: Make line-up with other lines.

**Pull request merged**